### PR TITLE
final proof is now sent on the common topic with META

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Reasoning behind the pull request
- send equivalent proofs on the common new topic
  
## Proposed changes
 - send equivalent proofs on the common new topic
 - small worker cleanup

## Testing procedure
- with feat branch

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: consensus/broadcast/metaChainMessenger.go [hunk 1]
```diff
@@ -112,6 +117,22 @@ func (mcm *metaChainMessenger) BroadcastHeader(header data.HeaderHandler, pkByte
...
+	if check.IfNilReflect(proof) {
+		return spos.ErrNilHeader
+	}
+
+	msgProof, err := mcm.marshalizer.Marshal(proof)
+	if err != nil {
+		return err
+	}
+
+	mcm.broadcast(common.EquivalentProofsTopic, msgProof, pkBytes)
```
Comment (AdoAdoAdo) [inline consensus/broadcast/metaChainMessenger.go:131]: is this a topic between metachain and all shards?

File: consensus/broadcast/delayedBroadcast.go [hunk 2]
```diff
@@ -245,37 +251,50 @@ func (dbb *delayedBlockBroadcaster) SetValidatorData(broadcastData *shared.Delay
 	return nil
 }
 
-// SetFinalConsensusMessageForValidator sets the consensus message to be broadcast by validator when its turn comes
-func (dbb *delayedBlockBroadcaster) SetFinalConsensusMessageForValidator(message *consensus.Message, consensusIndex int) error {
-	if message == nil {
-		return spos.ErrNilConsensusMessage
+// SetFinalProofForValidator sets the header proof to be broadcast by validator when its turn comes
+func (dbb *delayedBlockBroadcaster) SetFinalProofForValidator(
```
Comment (AdoAdoAdo) [inline consensus/broadcast/delayedBroadcast.go:255]: Where is the previous final message handled?
before equivalent proofs activation?

File: consensus/broadcast/delayedBroadcast.go [hunk 3]
```diff
@@ -808,34 +829,34 @@ func (dbb *delayedBlockBroadcaster) getBroadcastDelayForIndex(index int) time.Du
 	return 0
 }
 
-func (dbb *delayedBlockBroadcaster) consensusMessageAlarmExpired(alarmID string) {
+func (dbb *delayedBlockBroadcaster) finalProofAlarmExpired(alarmID string) {
```
Comment (AdoAdoAdo) [inline consensus/broadcast/delayedBroadcast.go:832]: same as my previous comment, where is the broadcast handled for the final info (before proofs activation)

File: consensus/broadcast/metaChainMessenger.go (hunk 1 above)
Comment (sstanculeanu) [reply consensus/broadcast/metaChainMessenger.go:131]: fixed

File: consensus/broadcast/delayedBroadcast.go (hunk 2 above)
Comment (sstanculeanu) [reply consensus/broadcast/delayedBroadcast.go:255]: these methods were added on purpose for equivalent proofs

File: consensus/broadcast/delayedBroadcast.go (hunk 3 above)
Comment (sstanculeanu) [reply consensus/broadcast/delayedBroadcast.go:832]: they were specific for equivalent proofs

File: consensus/broadcast/metaChainMessenger.go [hunk 4]
```diff
@@ -112,6 +117,22 @@ func (mcm *metaChainMessenger) BroadcastHeader(header data.HeaderHandler, pkByte
 	return nil
 }
 
+// BroadcastEquivalentProof will broadcast the proof for a header on the metachain common topic
+func (mcm *metaChainMessenger) BroadcastEquivalentProof(proof *block.HeaderProof, pkBytes []byte) error {
+	if check.IfNilReflect(proof) {
+		return spos.ErrNilHeader
```
Comment (ssd04) [inline consensus/broadcast/metaChainMessenger.go:123]: change error here

File: consensus/broadcast/shardChainMessenger.go [hunk 5]
```diff
@@ -123,6 +128,23 @@ func (scm *shardChainMessenger) BroadcastHeader(header data.HeaderHandler, pkByt
 	return nil
 }
 
+// BroadcastEquivalentProof will broadcast the proof for a header on the shard metachain common topic
+func (scm *shardChainMessenger) BroadcastEquivalentProof(proof *block.HeaderProof, pkBytes []byte) error {
+	if check.IfNilReflect(proof) {
+		return spos.ErrNilHeader
```
Comment (ssd04) [inline consensus/broadcast/shardChainMessenger.go:134]: same here

Comment (ssd04) [reply consensus/broadcast/shardChainMessenger.go:134]: maybe we can have a common inner function here for shard and meta and pass the topic

File: consensus/broadcast/metaChainMessenger.go (hunk 4 above)
Comment (sstanculeanu) [reply consensus/broadcast/metaChainMessenger.go:123]: updated

File: consensus/broadcast/shardChainMessenger.go (hunk 5 above)
Comment (sstanculeanu) [reply consensus/broadcast/shardChainMessenger.go:134]: fixed

Comment (sstanculeanu) [reply consensus/broadcast/shardChainMessenger.go:134]: applied the suggestion

**Review by AdoAdoAdo — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**